Use std::span more in StringImpl code

cdumez opened the pull request:
#### 27a1ee5ab6deb23c6e618ab2756faccce741e9e2
<pre>
Use std::span more in StringImpl code
<a href="https://bugs.webkit.org/show_bug.cgi?id=272208">https://bugs.webkit.org/show_bug.cgi?id=272208</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/dfg/DFGLazyJSValue.cpp:
(JSC::DFG::LazyJSValue::dumpInContext const):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::languageTagForLocaleID):
(JSC::addScriptlessLocaleIfNeeded):
(JSC::numberingSystemsForLocale):
(JSC::intlAvailableCalendars):
(JSC::intlAvailableTimeZones):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData):
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
* Source/WTF/wtf/HexNumber.h:
(WTF::StringTypeAdapter&lt;HexNumberBuffer&gt;::writeTo const):
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::escapeUnsafeCharacters):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::appendCharacters):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::allocateBuffer):
* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::find8NonASCIIAlignedImpl):
(WTF::find16NonASCIIAlignedImpl):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringConcatenate.h:
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::FormattedNumber::span const):
(WTF::StringTypeAdapter&lt;FormattedNumber&gt;::writeTo const):
(WTF::FormattedCSSNumber::span const):
(WTF::StringTypeAdapter&lt;FormattedCSSNumber&gt;::writeTo const):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::createInternal):
(WTF::StringImpl::createStaticStringImpl):
(WTF::StringImpl::create8BitIfPossible):
(WTF::StringImpl::foldCase):
(WTF::StringImpl::convertASCIICase):
(WTF::StringImpl::replace):
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::utf8ForCharactersIntoBuffer):
(WTF::StringImpl::tryGetUTF8 const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::create8BitIfPossible):
(WTF::StringImpl::createStaticStringImpl):
(WTF::StringImpl::copyCharacters):
(WTF::StringImpl::StringImpl):
(WTF::StringImpl::removeCharactersImpl):
(WTF::StringImpl::createByReplacingInCharacters):
(WTF::StringImpl::tryGetUTF8 const):
(WTF::StringImpl::tryGetUTF8ForCharacters):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::getCharacters8 const):
(WTF::StringView::getCharacters16 const):
(WTF::StringView::UpconvertedCharactersWithSize&lt;N&gt;::UpconvertedCharactersWithSize):
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::String::convertTo16Bit):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Access):
(WTF::uTextLatin1Extract):
(WTF::textLatin1ContextAwareMoveInPrimaryContext):
* Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:
(TestWebKitAPI::TEST(WTF_StringCommon, Find8NonASCII)):
(TestWebKitAPI::TEST(WTF_StringCommon, Find16NonASCII)):
* Tools/TestWebKitAPI/Tests/WTF/StringImpl.cpp:
(TestWebKitAPI::TEST(WTF, DynamicStaticStringImpl)):

Canonical link: <a href="https://commits.webkit.org/277147@main">https://commits.webkit.org/277147@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a51cd720165a3f7b1ebed7d3e8d1f7a713bde87e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26004 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49468 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49520 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42890 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49149 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30420 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23470 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47422 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/23012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4888 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/40090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/41847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51394 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46325 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18198 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23139 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44405 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/53467 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6562 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/22849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10983 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -273,13 +272,14 @@ class StringImpl : private StringImplShape {
     {
-        ASSERT(charactersAreAllASCII(std::span { bitwise_cast<const LChar*>(characters), static_cast<size_t>(length) }));
-        return createStaticStringImpl(bitwise_cast<const LChar*>(characters), length);
+        std::span lcharSpan { reinterpret_cast<const LChar*>(characters.data()), characters.size() };
```
Not `spanReinterpretCast`? Is it a “too many includes” issue?

cdumez replied:
That's right, this header doesn't currently include wtf/Algorithm.h and I was worried about increase build times. That said, I've been thinking that maybe `spanReinterpretCast` should be in wtf/StdLibExtras.h instead, which *is* included in Stringimpl.h.

cdumez replied:
Oh now that I made the change and let EWS chew on it, I remember why I didn't do this yet...

```
FAILED: Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o 
/usr/bin/ccache /usr/bin/c++ -DBUILDING_WEBKIT=1 -DBUILDING_WITH_CMAKE=1 -DBUILDING_WPE__=1 -DBWRAP_EXECUTABLE=\"/usr/bin/bwrap\" -DDBUS_PROXY_EXECUTABLE=\"/usr/bin/xdg-dbus-proxy\" -DGETTEXT_PACKAGE=\"WPE\" -DHAVE_CONFIG_H=1 -DJSC_GLIB_API_ENABLED -DPAS_BMALLOC=1 -DSKIA_IMPLEMENTATION=1 -DSK_ASSUME_GL=0 -DSK_ASSUME_GL_ES=1 -DSK_ASSUME_WEBGL=0 -DSK_CODEC_DECODES_PNG -DSK_DISABLE_LEGACY_GL_MAKE_NATIVE_INTERFACE -DSK_DISABLE_LEGACY_IMAGE_READBUFFER -DSK_DISABLE_LEGACY_INIT_DECODERS -DSK_DISABLE_LEGACY_PNG_WRITEBUFFER -DSK_DISABLE_TRACING -DSK_DISABLE_WEBGL_INTERFACE -DSK_ENABLE_AVX512_OPTS -DSK_ENABLE_PRECOMPILE -DSK_GAMMA_APPLY_TO_A8 -DSK_GANESH -DSK_GL -DSK_R32_SHIFT=16 -DSK_RELEASE -DSK_TYPEFACE_FACTORY_FREETYPE -DSTATICALLY_LINKED_WITH_WTF -DSTATICALLY_LINKED_WITH_bmalloc -I/app/webkit/Source/WTF -I/app/webkit/WebKitBuild/WPE/Release -I/app/webkit/Source/ThirdParty/skia -I/app/webkit/WebKitBuild/WPE/Release/WTF/Headers -I/app/webkit/WebKitBuild/WPE/Release/bmalloc/Headers -isystem /usr/include/freetype2 -fdiagnostics-color=always -Wextra -Wall -pipe -fmax-errors=20 -Wno-odr -Wno-stringop-overread -Wno-stringop-overflow -Wno-nonnull -Wno-array-bounds -Wno-expansion-to-defined -Wno-noexcept-type -Wno-psabi -Wno-misleading-indentation -Wno-maybe-uninitialized -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fno-strict-aliasing -fno-exceptions -fno-rtti -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++17 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -ffp-contract=off -fstrict-aliasing -Wno-attributes -Wno-cast-align -Wno-dangling-reference -Wno-deprecated -Wno-psabi -Wno-suggest-attribute=format -Wno-undef -Wno-uninitialized -Wno-unused-parameter -MD -MT Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o -MF Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o.d -o Source/ThirdParty/skia/CMakeFiles/Skia.dir/__/__/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp.o -c /app/webkit/Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp
In file included from /app/webkit/Source/WTF/wtf/FastMalloc.h:26,
                 from /app/webkit/Source/WTF/config.h:31,
                 from /app/webkit/Source/WebCore/platform/skia/SkiaAllocatorFastMalloc.cpp:26:
/app/webkit/Source/WTF/wtf/StdLibExtras.h:675:6: error: ‘span’ in namespace ‘std’ does not name a template type
  675 | std::span<T> spanReinterpretCast(std::span<U> span)
      |      ^~~~
/app/webkit/Source/WTF/wtf/StdLibExtras.h:675:1: note: ‘std::span’ is only available from C++20 onwards
  675 | std::span<T> spanReinterpretCast(std::span<U> span)
      | ^~~
/app/webkit/Source/WTF/wtf/StdLibExtras.h:733:12: error: ‘spanReinterpretCast’ has not been declared in ‘WTF’
```

The Skia bot is importing this header in some of their Skia logic and they're not building with C++20 support enabled (so no std::span support) :(

darinadler replied:
We must fix the build so it compiles those source files with C++20 support enabled; not acceptable to compile our headers without it

darinadler replied:
Is that challenging?

cdumez replied:
I have made people maintaining the skia port aware of the issue. I do agree they should build with C++20 support enabled. In the mean time, I did address the bot failure by adding a C++ version guard around the new function in the header.

cdumez replied:
Oh, it looks like it is still broken for Igalia's JSC bots though :(
```
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h: In static member function ‘static WTF::Ref<WTF::StringImpl> WTF::StringImpl::create(std::span<const char>)’:
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h:250:99: error: ‘spanReinterpretCast’ was not declared in this scope
  250 |     ALWAYS_INLINE static Ref<StringImpl> create(std::span<const char> characters) { return create(spanReinterpretCast<const LChar>(characters)); }
      |                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h: In static member function ‘static WTF::Ref<WTF::StringImpl> WTF::StringImpl::createWithoutCopying(std::span<const char>)’:
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h:266:127: error: ‘spanReinterpretCast’ was not declared in this scope
  266 |     ALWAYS_INLINE static Ref<StringImpl> createWithoutCopying(std::span<const char> characters) { return createWithoutCopying(spanReinterpretCast<const LChar>(characters)); }
      |                                                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h: In static member function ‘static WTF::Ref<WTF::StringImpl> WTF::StringImpl::createStaticStringImpl(std::span<const char>)’:
/home/buildbot/igalia-jsc32-armv7-ews/JSC-ARMv7-32bits-Build-EWS/build/Source/WTF/wtf/text/StringImpl.h:277:31: error: ‘spanReinterpretCast’ was not declared in this scope
  277 |         std::span lcharSpan = spanReinterpretCast<const LChar>(characters);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[3]: *** [Source/WTF/wtf/CMakeFiles/WTF.dir/build.make:186: Source/WTF/wtf/CMakeFiles/WTF.dir/CodePtr.cpp.o] Error 1
```

cdumez replied:
I ping'd Igalia about this as well. But due to timezone differences, I may need to go back to my original version until theirs ports consistently build with C++20 support enabled (it was not directly obvious to me how to make that change for their ports).

cdumez replied:
`#if __cplusplus >= 202002L` doesn't evaluate to true on those Igalia JSC bots but they can build span code though...

cdumez committed:
Use std::span more in StringImpl code
https://bugs.webkit.org/show_bug.cgi?id=272208

Reviewed by Darin Adler.

* Source/JavaScriptCore/dfg/DFGLazyJSValue.cpp:
(JSC::DFG::LazyJSValue::dumpInContext const):
* Source/JavaScriptCore/runtime/IntlObject.cpp:
(JSC::languageTagForLocaleID):
(JSC::addScriptlessLocaleIfNeeded):
(JSC::numberingSystemsForLocale):
(JSC::intlAvailableCalendars):
(JSC::intlAvailableTimeZones):
* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData):
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings):
* Source/WTF/wtf/HexNumber.h:
(WTF::StringTypeAdapter<HexNumberBuffer>::writeTo const):
* Source/WTF/wtf/URLHelpers.cpp:
(WTF::URLHelpers::escapeUnsafeCharacters):
* Source/WTF/wtf/text/StringBuilder.cpp:
(WTF::StringBuilder::appendCharacters):
* Source/WTF/wtf/text/StringBuilderInternals.h:
(WTF::StringBuilder::allocateBuffer):
* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::find8NonASCIIAlignedImpl):
(WTF::find16NonASCIIAlignedImpl):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringConcatenate.h:
* Source/WTF/wtf/text/StringConcatenateNumbers.h:
(WTF::FormattedNumber::span const):
(WTF::StringTypeAdapter<FormattedNumber>::writeTo const):
(WTF::FormattedCSSNumber::span const):
(WTF::StringTypeAdapter<FormattedCSSNumber>::writeTo const):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::createInternal):
(WTF::StringImpl::createStaticStringImpl):
(WTF::StringImpl::create8BitIfPossible):
(WTF::StringImpl::foldCase):
(WTF::StringImpl::convertASCIICase):
(WTF::StringImpl::replace):
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::utf8ForCharactersIntoBuffer):
(WTF::StringImpl::tryGetUTF8 const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::create8BitIfPossible):
(WTF::StringImpl::createStaticStringImpl):
(WTF::StringImpl::copyCharacters):
(WTF::StringImpl::StringImpl):
(WTF::StringImpl::removeCharactersImpl):
(WTF::StringImpl::createByReplacingInCharacters):
(WTF::StringImpl::tryGetUTF8 const):
(WTF::StringImpl::tryGetUTF8ForCharacters):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::getCharacters8 const):
(WTF::StringView::getCharacters16 const):
(WTF::StringView::UpconvertedCharactersWithSize<N>::UpconvertedCharactersWithSize):
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::make8Bit):
(WTF::String::convertTo16Bit):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Access):
(WTF::uTextLatin1Extract):
(WTF::textLatin1ContextAwareMoveInPrimaryContext):
* Tools/TestWebKitAPI/Tests/WTF/StringCommon.cpp:
(TestWebKitAPI::TEST(WTF_StringCommon, Find8NonASCII)):
(TestWebKitAPI::TEST(WTF_StringCommon, Find16NonASCII)):
* Tools/TestWebKitAPI/Tests/WTF/StringImpl.cpp:
(TestWebKitAPI::TEST(WTF, DynamicStaticStringImpl)):

Canonical link: https://commits.webkit.org/277147@main
